src/base/init/init.c

```diff
@@ -405,15 +403,38 @@ void low_mem_init(void)
   }
 
   if (config.ext_mem) {
-    /* LOWMEM_SIZE + HMASIZE == base
-     * Use dpmi_rsv_low as ext_mem. */
-    register_hardware_ram_virtual('X', LOWMEM_SIZE + HMASIZE, EXTMEM_SIZE,
-	    MEM_BASE32(LOWMEM_SIZE + HMASIZE), LOWMEM_SIZE + HMASIZE);
+    /* LOWMEM_SIZE + HMASIZE == base */
     memcheck_addtype('X', "EXT MEM");
     memcheck_reserve('X', LOWMEM_SIZE + HMASIZE, EXTMEM_SIZE);
     x_printf("Ext.Mem of size 0x%x at %#x\n", EXTMEM_SIZE, result + LOWMEM_SIZE + HMASIZE);
   }
 
+  ptr += phys_rsv + dpmi_rsv_low + dpmi_mem_size();
+  /* create non-identity mapping up to dpmi_base */
+  phys_rsv = config.dpmi_base - (LOWMEM_SIZE + HMASIZE);
+  ptr2 = smalloc_aligned_topdown(&main_pool, MEM_BASE32(memsize),
+	PAGE_SIZE, phys_rsv);
+  assert(ptr2);
+  /* establish alias access for int15 */
+  register_hardware_ram_virtual('X', LOWMEM_SIZE + HMASIZE, phys_rsv,
+	    lowmem + LOWMEM_SIZE + HMASIZE, DOSADDR_REL(ptr2));
```

Comment:
While working on aligned allocs I just realized this is not correct: `lowmem + LOWMEM_SIZE + HMASIZE` is only valid memory up to EXTMEM_SIZE, the rest is random shm. So a program may lock XMS, write to it, but end up writing to VGAEMU memory!

you could use `ptr2` instead to map to the aliased RAM in high linear DPMI space, or, another valid thing would be to do:
```
  register_hardware_ram_virtual('X', LOWMEM_SIZE + HMASIZE, EXTMEM_SIZE,
	    lowmem + LOWMEM_SIZE + HMASIZE, DOSADDR_REL(ptr2));
  register_hardware_ram_virtual('x', LOWMEM_SIZE + HMASIZE + EXTMEM_SIZE, phys_rsv - EXTMEM_SIZE,
	    ptr2 + EXTMEM_SIZE, DOSADDR_REL(ptr2) + EXTMEM_SIZE);
```

Reply:
That's a design bug actually. :(
Had to remove that arg after adding
aliasmaps because alias_mapping_pa()
now updates it.

Reply:
The issue is now that you've also moved the vgaemu LFB mapping in KVM, it now maps also maps `0xe0000000` to `vga.mem.lfb_base = &mem_base[0xa5f6000]` (default values) instead of `vga.mem.base` which is in SHM.

`alias_mapping_pa` could solve that, but needs to be careful as it's not only called at init time.

Reply:
An advantage of `alias_mapping_pa` calling `mmap_kvm` after init time (from `map_EMB()`) though it that then dosemu's idea of physical memory is in sync with KVM. It just means the slots are not fixed at init time any more but adjusted on the fly once again, so it's more work for KVM.

In the end though this PR isn't broken, you can happily map the VGAEMU LFB to DPMI lin space as mprotect doesn't interfere, though again this isn't huge page aligned now for different reasons. One can either merge this as is and address the rest in future PRs or keep working on this.

Reply:
or maybe just for now, split off the `mmap_kvm()` call from `register_hardware_ram_virtual()` again and deal with the rest later?

Reply:
So where is the problem?
It maps lfb to &mem_base[0xa5f6000]
and then alias_mapping_pa() puts shm
there. KVM gets the alias instead of
"directly to source", but what is the problem?
Yes, its unaligned, but use smalloc_aligned()
then?

Reply:
OK if that's a problem, lets add a separate
func for lfb. Done now.

Reply:
thanks, ok, looks good to me!